### ISSUE TITLE
[ARM] clang/LLVM silently switches to the soft-float ABI if an FPU-less CPU is specified for a hard-float target

### ISSUE THREAD

**Issue by chrisnc** [llvm/llvm-project#110383]
In GCC, this is treated as an error, because the chosen CPU cannot implement the ABI that has been requested.

https://godbolt.org/z/89YGKW79P

**Comment by RalfJung** [comment]
x86 has a similar issue, where `-x87` on a typical hard-float target will silently switch to the soft-float ABI. IMO it'd be better to error there, too, so that an explicit `+soft-float` is required to use the soft-float ABI -- that avoids nasty surprises where code doesn't use the ABI it is expected to use.

**Comment by RalfJung** [comment]
Yeah, that makes sense. Ideally it would be an error in llvm as well, in a way that the logic does not live in clang but can be reused by other frontends.

**Comment by RalfJung** [comment]
A similar issue exists with target features: if I pass `-fpregs` to `createTargetMachine`, it will silently change the ABI. I would expect an error saying that the registers are missing for this ABI. If I pass `+soft-float,-fpregs`, then everything is fine as I explicitly asked for the ABI change.

**Comment by chrisnc** [comment]
@RalfJung I believe the PR I have linked to this issue https://github.com/llvm/llvm-project/pull/111334 will address that case as well. If the hard-float ABI is used but the fpregs feature isn't present, it errors out. I'll confirm this later.

**Comment by RalfJung** [comment]
That would be great :)
